File: skills/search/scripts/update_job.py
#!/usr/bin/env python3
"""Update a pipeline role in jobs.md: move stage, dismiss, or restore.
The ONE structured write for user actions arriving via chat.

Fails loudly on 0 or 2+ matches — an ambiguous request never silently edits
the wrong role. (Ported from the retired board skill, 2026-08-14; storage is
now the readable jobs.md record.)

  python3 update_job.py --workspace . --company Cursor --title "Forward Deployed" --stage Applied
  python3 update_job.py --workspace . --company Anthropic --title "Continuous Deployment" --dismiss --reason "internal CI/CD, not FDE"
  python3 update_job.py --workspace . --company Anthropic --title "..." --restore
"""
import argparse, sys
import jobs_md as jm


def main():
    p = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument("--workspace", required=True)
    p.add_argument("--company", required=True)
    p.add_argument("--title", required=True, help="full or partial — must match exactly one role")
    g = p.add_mutually_exclusive_group(required=True)
    g.add_argument("--stage", choices=jm.STAGES)
    g.add_argument("--dismiss", action="store_true")
    g.add_argument("--restore", action="store_true")
    p.add_argument("--reason", help="dismiss reason (with --dismiss)")
    a = p.parse_args()

    rows = jm.load(a.workspace)
    hits = jm.find(rows, a.company, a.title)
    if len(hits) != 1:
        which = ("no roles" if not hits else
                 f"{len(hits)} roles: " + "; ".join(r["title"] for r in hits))
        print(f"error: --company/--title matched {which} — be more specific", file=sys.stderr)
        return 2
    r = hits[0]
    r["updated_at"] = jm.now_iso()
    if a.stage:
        r["stage"], r["dismissed"] = a.stage, False
        action = f"stage -> {a.stage}"
    elif a.dismiss:
        r["dismissed"] = True
        if a.reason:
            r["dismiss_reason"] = a.reason
        action = "dismissed" + (f" ({a.reason})" if a.reason else "")
    else:
        r["dismissed"] = False
        action = "restored"
    jm.save(a.workspace, rows)
    print(f"updated: {r['company']} — {r['title']} — {action}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: packages/checkers/src/update-job.mjs
// A faithful JS port of skills/search/scripts/update_job.py. Runs in Node
// AND in the browser; no node:fs, no node:path.
import * as jm from "./jobs-md.mjs";
import { parseFlags, argError } from "./argx.mjs";

const PROG = "update_job.py";
const USAGE =
  "usage: update_job.py [-h] --workspace WORKSPACE --company COMPANY\n" +
  "                     --title TITLE\n" +
  "                     (--stage {To Review,Interested,Applied,Interviewing,Offer} |\n" +
  "                     --dismiss | --restore) [--reason REASON]\n";
const OPTIONS = [
  { flag: "--workspace", dest: "workspace", required: true },
  { flag: "--company", dest: "company", required: true },
  { flag: "--title", dest: "title", required: true },
  { flag: "--stage", dest: "stage", choices: jm.STAGES },
  { flag: "--dismiss", dest: "dismiss", boolean: true },
  { flag: "--restore", dest: "restore", boolean: true },
  { flag: "--reason", dest: "reason" },
];

export async function run(argv, io, now = () => new Date()) {
  const parsed = parseFlags(argv, { options: OPTIONS });
  if (parsed.error) return argError(PROG, USAGE, parsed.error);
  const a = parsed.args;

  // argparse's mutually-exclusive group: exactly one of --stage/--dismiss/
  // --restore, checked (and reported) in the order the flags appear on argv.
  const order = argv.map((t) => t.split("=")[0]).filter((t) => t === "--stage" || t === "--dismiss" || t === "--restore");
  if (order.length === 0) {
    return argError(PROG, USAGE, "one of the arguments --stage --dismiss --restore is required");
  }
  if (order.length > 1) {
    return argError(PROG, USAGE, `argument ${order[1]}: not allowed with argument ${order[0]}`);
  }

  const rows = await jm.load(io, a.workspace);
  const hits = jm.find(rows, a.company, a.title);
  if (hits.length !== 1) {
    const which = hits.length === 0 ? "no roles" : `${hits.length} roles: ` + hits.map((r) => r.title).join("; ");
    return { stdout: "", stderr: `error: --company/--title matched ${which} — be more specific\n`, exitCode: 2 };
  }
  const r = hits[0];
  r.updated_at = jm.nowIso(now);
  let action;
  if (a.stage) {
    r.stage = a.stage;
    r.dismissed = false;
    action = `stage -> ${a.stage}`;
  } else if (a.dismiss) {
    r.dismissed = true;
    if (a.reason) r.dismiss_reason = a.reason;
    action = "dismissed" + (a.reason ? ` (${a.reason})` : "");
  } else {
    r.dismissed = false;
    action = "restored";
  }
  try {
    await jm.save(io, a.workspace, rows, { now });
  } catch (e) {
    if (e instanceof jm.DuplicateKeyError) return { stdout: "", stderr: e.message + "\n", exitCode: 1 };
    throw e;
  }
  return { stdout: `updated: ${r.company} — ${r.title} — ${action}\n`, stderr: "", exitCode: 0 };
}
